server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -156,29 +162,27 @@ public void sendPublishRequest(DiscoveryNode destination, PublishRequest publish
                                            ActionListener<PublishWithJoinResponse> responseActionListener) {
                 assert publishRequest.getAcceptedState() == clusterChangedEvent.state() : "state got switched on us";
                 if (destination.equals(nodes.getLocalNode())) {
-                    // the master needs the original non-serialized state as the cluster state contains some volatile information that we
-                    // don't want to be replicated because it's not usable on another node (e.g. UnassignedInfo.unassignedTimeNanos) or
-                    // because it's mostly just debugging info that would unnecessarily blow up CS updates (I think there was one in
-                    // snapshot code).
-                    // TODO: look into these and check how to get rid of them
-                    transportService.getThreadPool().generic().execute(new AbstractRunnable() {
+                    // if publishing to self, use original request instead (see currentPublishRequestToSelf for explanation)
+                    final PublishRequest previousRequest = currentPublishRequestToSelf.getAndSet(publishRequest);
+                    assert previousRequest == null;
+                    final ActionListener<PublishWithJoinResponse> originalListener = responseActionListener;
+                    responseActionListener = new ActionListener<PublishWithJoinResponse>() {
```

Comment:
I'd rather this used a different variable rather than mutating the parameter.

Reply:
fixed in 9c2822d